pkgs/development/haskell-modules/generic-builder.nix

```diff
@@ -383,11 +383,38 @@ stdenv.mkDerivation ({
       mv tmp "$d"
     done
 
-    for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
-      for lib in "$d/"*.{dylib,so}; do
-        ln -s "$lib" "$dynamicLinksDir"
+    # This function makes the links of shared libraries this package uses,
+    # and it avoids creating duplicate links for the same file.
+    # $packageConfDir has conf files for each haskell package this package uses.
+    # There are the links of shared libraries in dynamic-library-dirs of $packageConfDir.
+    # The links are duplicated in some haskell packages.
+    makeDynamicLinks () {
+      for d in $(grep '^dynamic-library-dirs:' "$packageConfDir"/* | cut -d' ' -f2- | tr ' ' '\n' | sort -u); do
+        for lib in "$d/"*.{dylib,so} ; do
+          # When '*.{dylib,so}' does not match anything, '*.{dylib,so}' becomes '*.dylib or *.so' including '*'
+          # [ -f "$lib" ] checks if the matched path is available.
```

Comment:
You can use `shopt -s nullglob` to prevent this from happening and needing this condition